File: koin/src/main/java/in/koreatech/koin/ui/login/LoginActivity.java
package in.koreatech.koin.ui.login;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import in.koreatech.koin.R;
import in.koreatech.koin.core.activity.ActivityBase;
import in.koreatech.koin.core.progressdialog.CustomProgressDialog;
import in.koreatech.koin.ui.main.MainActivity;
import in.koreatech.koin.ui.signup.SignupActivity;
import in.koreatech.koin.ui.login.presenter.LoginContract;
import in.koreatech.koin.ui.login.presenter.LoginPresenter;
import in.koreatech.koin.core.toast.ToastUtil;
import in.koreatech.koin.ui.forgotpassword.ForgotPasswordActivity;

import static in.koreatech.koin.util.FormValidatorUtil.validateStringIsEmpty;

public class LoginActivity extends ActivityBase implements LoginContract.View {
    private final String TAG = "LoginActivity";
    private Context context;
    private LoginPresenter loginPresenter;
    private boolean isMainActivity;

    /* View Components */
    @BindView(R.id.login_edittext_id)
    EditText editTextID;
    @BindView(R.id.login_edittext_pw)
    EditText editTextPW;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        this.isMainActivity = getIntent().getBooleanExtra("FIRST_LOGIN", true);
        ButterKnife.bind(this);
        init();
    }

    private void init() {
        context = this;
        new LoginPresenter(this);
    }

    @Override
    public void setPresenter(LoginPresenter presenter) {
        this.loginPresenter = presenter;
    }

    @OnEditorAction({R.id.login_edittext_id, R.id.login_edittext_pw})
    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
        switch (v.getId()) {
            case R.id.login_edittext_id:
                this.editTextPW.requestFocus();
                break;
            case R.id.login_edittext_pw:
                View view = this.getCurrentFocus();
                if (view != null) {
                    InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
                    Objects.requireNonNull(imm).hideSoftInputFromWindow(view.getWindowToken(), 0);
                }
                break;
            default:
                break;
        }
        return false;
    }

    /**
     * 로그인버튼 onclick 메소드
     * 키보드 숨김
     * 아이디/비밀번호 입력 여부 확인
     * LoginPresenter의 login 메소드 호출(password hash false)
     */
    @OnClick(R.id.login_button)
    public void onClickLoginButton() {
        View view = this.getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            Objects.requireNonNull(imm).hideSoftInputFromWindow(view.getWindowToken(), 0);
        }

        if (validateStringIsEmpty(this.editTextID.getText().toString()) || validateStringIsEmpty(this.editTextPW.getText().toString())) {
            ToastUtil.getInstance().makeShort("아이디와 비밀번호를 입력해주세요");
        } else {
            this.loginPresenter.login(this.editTextID.getText().toString(), this.editTextPW.getText().toString(), false);
        }
    }

    @OnClick(R.id.login_button_signup)
    public void onClickSignUpButton() {
        startActivity(new Intent(LoginActivity.this, SignupActivity.class));
    }

    @OnClick(R.id.forgot_password_linearLayout)
    public void onClickForgotPassword() {
        startActivity(new Intent(LoginActivity.this, ForgotPasswordActivity.class));
    }

    @OnClick(R.id.anonymous_login_linearLayout)
    public void onClickAnonymousLogin() {
        if (this.isMainActivity) {
            startActivity(new Intent(LoginActivity.this, MainActivity.class));
            this.isMainActivity = false;
        }
        finish();
    }

    @Override
    public void gotoMain() {
        if (this.isMainActivity) {
            startActivity(new Intent(LoginActivity.this, MainActivity.class));
            this.isMainActivity = false;
        }
        finish();
    }

    @Override
    public void showProgress() {
        showProgressDialog("진행 중");
    }

    @Override
    public void hideProgress() {
        hideProgressDialog();
    }

    /**
     * Toast 메시지 띄우는 메소드
     *
     * @param message toast 메시지 내용
     */
    @Override
    public void showMessage(String message) {
        ToastUtil.getInstance().makeShort(message);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }
}

File: koin/src/main/java/in/koreatech/koin/ui/login/LoginActivity.kt
package `in`.koreatech.koin.ui.login

import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP
import android.net.Uri
import android.os.Bundle
import android.view.MotionEvent
import androidx.activity.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import dagger.hilt.android.AndroidEntryPoint
import `in`.koreatech.koin.R
import `in`.koreatech.koin.common.UiStatus
import `in`.koreatech.koin.core.activity.ActivityBase
import `in`.koreatech.koin.core.analytics.EventAction
import `in`.koreatech.koin.core.analytics.EventLogger
import `in`.koreatech.koin.core.constant.AnalyticsConstant
import `in`.koreatech.koin.core.util.dataBinding
import `in`.koreatech.koin.databinding.ActivityLoginBinding
import `in`.koreatech.koin.ui.businesslogin.BusinessLoginActivity
import `in`.koreatech.koin.ui.forgotpassword.ForgotPasswordActivity
import `in`.koreatech.koin.ui.login.viewmodel.LoginViewModel
import `in`.koreatech.koin.ui.main.activity.MainActivity
import `in`.koreatech.koin.ui.signup.SignupActivity
import `in`.koreatech.koin.util.SnackbarUtil
import `in`.koreatech.koin.util.ext.hideKeyboard
import `in`.koreatech.koin.util.ext.textString
import `in`.koreatech.koin.util.ext.withLoading
import kotlinx.coroutines.launch

@AndroidEntryPoint
class LoginActivity : ActivityBase(R.layout.activity_login) {
    private val binding by dataBinding<ActivityLoginBinding>()
    override val screenTitle = "로그인"

    private val loginViewModel by viewModels<LoginViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initView()
        initViewModel()
    }

    private fun initViewModel() = with(loginViewModel) {
        withLoading(this@LoginActivity, this)

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                loginState.collect {
                    when (it.status) {
                        is UiStatus.Init -> Unit
                        is UiStatus.Loading -> Unit
                        is UiStatus.Success -> goToNextRoute()
                        is UiStatus.Failed -> {
                            SnackbarUtil.makeShortSnackbar(binding.root, it.status.message)
                            loginViewModel.onFailedLogin()
                        }
                    }
                }
            }
        }
    }

    private fun goToNextRoute() {
        val uri = intent.data
        val link = uri?.getQueryParameter("link")

        if (link != null) {
            try {
                val intent = Intent(Intent.ACTION_VIEW).apply {
                    flags = FLAG_ACTIVITY_CLEAR_TOP
                    data = Uri.parse(link)
                }
                startActivity(intent)
                finish()
            } catch (e: Exception) {
                startActivity(Intent(this@LoginActivity, MainActivity::class.java))
                finish()
            }
        } else {
            startActivity(Intent(this@LoginActivity, MainActivity::class.java))
            finish()
        }
    }

    private fun initView() = with(binding) {
        loginEdittextId.setOnEditorActionListener { v, actionId, event ->
            loginEdittextPw.requestFocus()
        }
        loginEdittextPw.setOnEditorActionListener { v, actionId, event ->
            currentFocus?.also { view ->
                hideKeyboard(view)
            }
            false
        }

        loginButton.setOnClickListener {
            currentFocus?.let { view ->
                hideKeyboard(view)
            }

            if (
                loginEdittextId.textString.isBlank() ||
                loginEdittextPw.textString.isBlank()
            ) {
                SnackbarUtil.makeShortSnackbar(
                    binding.root,
                    getString(R.string.login_required_field_not_filled)
                )
            } else {
                loginViewModel.login(
                    email = getString(
                        R.string.koreatech_email_postfix,
                        loginEdittextId.text.trim()
                    ),
                    password = loginEdittextPw.text.toString().trim()
                )
            }
            EventLogger.logClickEvent(
                EventAction.USER,
                AnalyticsConstant.Label.LOGIN,
                getString(R.string.login_complete)
            )
        }

        loginButtonSignup.setOnClickListener {
            startActivity(Intent(this@LoginActivity, SignupActivity::class.java))
            EventLogger.logClickEvent(
                EventAction.USER,
                AnalyticsConstant.Label.LOGIN,
                getString(R.string.sign_up)
            )
        }

        forgotPasswordLinearLayout.setOnClickListener {
            EventLogger.logClickEvent(
                EventAction.USER,
                AnalyticsConstant.Label.LOGIN,
                getString(R.string.find_password)
            )
            startActivity(Intent(this@LoginActivity, ForgotPasswordActivity::class.java))
        }

        anonymousLoginLinearLayout.setOnClickListener {
            startActivity(Intent(this@LoginActivity, MainActivity::class.java))
            finish()
        }

        isBusinessButton.setOnClickListener {
            startActivity(Intent(this@LoginActivity, BusinessLoginActivity::class.java))
        }
    }

    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
        hideKeyboard(currentFocus ?: binding.root)
        return super.dispatchTouchEvent(ev)
    }
}